# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationParent.cpp
# Commit: c942992fa981
# Full Hash: c942992fa981c66c424f9e16a50977529d4f6c54
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2024-11-12 09:32:50
# Regressor Bug: 1928702
# File Overlap Count: 1
# Description:
#   Bug 1928702 - Part 10: Implement alertfinished behavior r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D227695
# ==============================================================================

diff -r 11a3c2c86a8e -r c942992fa981 dom/notification/NotificationParent.cpp
--- a/dom/notification/NotificationParent.cpp	Mon Nov 11 20:34:14 2024 +0000
+++ b/dom/notification/NotificationParent.cpp	Mon Nov 11 20:34:15 2024 +0000
@@ -6,7 +6,9 @@
 
 #include "NotificationParent.h"
 
+#include "nsThreadUtils.h"
 #include "NotificationUtils.h"
+#include "mozilla/StaticPrefs_dom.h"
 #include "mozilla/ipc/Endpoint.h"
 #include "mozilla/Components.h"
 #include "nsComponentManagerUtils.h"
@@ -56,7 +58,17 @@
       mResolver.take().value()(CopyableErrorResult(NS_ERROR_FAILURE));
       return NS_OK;
     }
+
+    // XXX: QM_TRY?
+    (void)NS_WARN_IF(NS_FAILED(UnpersistNotification(mPrincipal, mId)));
+    (void)NS_WARN_IF(NS_FAILED(FireCloseEvent()));
+    Close();
+
+    return NS_OK;
   }
+
+  MOZ_ASSERT_UNREACHABLE("Unknown notification topic");
+
   return NS_OK;
 }
 
@@ -84,6 +96,28 @@
   return NS_ERROR_FAILURE;
 }
 
+nsresult NotificationParent::FireCloseEvent() {
+  // This needs to be done here rather than in the child actor's
+  // RecvNotifyClose because the caller might not be in the service worker
+  // context but in the window context
+  if (nsCOMPtr<nsIServiceWorkerManager> swm =
+          mozilla::components::ServiceWorkerManager::Service()) {
+    nsAutoCString originSuffix;
+    MOZ_TRY(mPrincipal->GetOriginSuffix(originSuffix));
+    nsAutoString behavior;
+    if (!mOptions.behavior().ToJSON(behavior)) {
+      return NS_ERROR_FAILURE;
+    }
+    MOZ_TRY(swm->SendNotificationCloseEvent(
+        originSuffix, NS_ConvertUTF16toUTF8(mScope), mId, mOptions.title(),
+        NS_ConvertASCIItoUTF16(GetEnumString(mOptions.dir())), mOptions.lang(),
+        mOptions.body(), mOptions.tag(), mOptions.icon(),
+        mOptions.dataSerialized(), behavior));
+    return NS_OK;
+  }
+  return NS_ERROR_FAILURE;
+}
+
 // Step 4 of
 // https://notifications.spec.whatwg.org/#dom-notification-notification
 mozilla::ipc::IPCResult NotificationParent::RecvShow(ShowResolver&& aResolver) {